components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketThriftClient.java

```diff
@@ -0,0 +1,234 @@
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TSSLTransportFactory;
+import org.apache.thrift.transport.TTransport;
+import org.apache.thrift.transport.TTransportException;
+import org.wso2.carbon.apimgt.api.dto.ConditionDTO;
+import org.wso2.carbon.apimgt.api.model.URITemplate;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityConstants;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityException;
+import org.wso2.carbon.apimgt.gateway.handlers.security.thrift.ThriftKeyValidatorClient;
+import org.wso2.carbon.apimgt.gateway.handlers.security.thrift.ThriftUtils;
+import org.wso2.carbon.apimgt.impl.dto.APIKeyValidationInfoDTO;
+import org.wso2.carbon.apimgt.impl.generated.thrift.APIKeyValidationService;
+import org.wso2.carbon.apimgt.impl.generated.thrift.ConditionGroupDTO;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class WebsocketThriftClient {
+	private static final Log log = LogFactory.getLog(ThriftKeyValidatorClient.class);
+	private ThriftUtils thriftUtils = null;
+	private String sessionId = null;
+	private APIKeyValidationService.Client keyValClient = null;
+
+	public WebsocketThriftClient() throws APISecurityException {
+		try {
+			thriftUtils = ThriftUtils.getInstance();
+			sessionId = thriftUtils.getSessionId();
+			//create new APIKeyValidator client
+			TSSLTransportFactory.TSSLTransportParameters param =
+					new TSSLTransportFactory.TSSLTransportParameters();
+
+			param.setTrustStore(thriftUtils.getTrustStorePath(),
+			                    thriftUtils.getTrustStorePassword());
+
+			TTransport transport = TSSLTransportFactory
+					.getClientSocket(ThriftUtils.getThriftServerHost(), thriftUtils.getThriftPort(),
+					                 thriftUtils.getThriftClientConnectionTimeOut(), param);
+
+			//TProtocol protocol = new TCompactProtocol(transport);
+			//TODO:needs to decide on the optimum protocol.
+			TProtocol protocol = new TBinaryProtocol(transport);
+
+			//create thrift based entitlement service client and invoke
+			keyValClient = new APIKeyValidationService.Client(protocol);
+
+		} catch (TTransportException e) {
+			log.error("Could not connect to Thrift host", e);
+			throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+			                               e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * Initialize thrift key validation
+	 *
+	 * @param context
+	 * @param apiVersion
+	 * @param apiKey
+	 * @return
+	 * @throws APISecurityException
+	 */
+	public APIKeyValidationInfoDTO getAPIKeyData(String context, String apiVersion, String apiKey)
+			throws APISecurityException {
+
+		APIKeyValidationInfoDTO apiKeyValidationInfoDTO = null;
+		org.wso2.carbon.apimgt.impl.generated.thrift.APIKeyValidationInfoDTO thriftDTO;
+
+		try {
+			thriftDTO =
+					keyValClient.validateKeyforHandshake(context, apiVersion, apiKey, sessionId);
+
+		} catch (Exception e) {
+			try {
+
```

Comment:
Its hard to avoid it here because this retry logic should be executed once an exception is cought